lib/FrontendTool/FrontendTool.cpp

```diff
@@ -132,6 +142,67 @@ static bool emitMakeDependencies(DiagnosticEngine &diags,
   return false;
 }
 
+static bool emitLoadedModuleTrace(ASTContext &ctxt,
+                                  DependencyTracker &depTracker,
+                                  const FrontendOptions &opts) {
+  std::error_code EC;
+  llvm::raw_fd_ostream out(opts.LoadedModuleTracePath, EC,
+                           llvm::sys::fs::F_None);
+
+  if (out.has_error() || EC) {
+    ctxt.Diags.diagnose(SourceLoc(), diag::error_opening_output,
+                        opts.LoadedModuleTracePath, EC.message());
+    out.clear_error();
+    return true;
+  }
+
+  using StringVector = llvm::SmallVector<std::string, 16>;
+  StringVector swiftModules;
+  StringVector apiNotes;
+
+  // Canonicalise all the paths by opening them.
+  for (auto &dep : depTracker.getDependencies()) {
+    llvm::SmallString<256> buffer;
+    StringRef realPath;
+    int FD;
+    // FIXME: appropriate error handling
+    if (llvm::sys::fs::openFileForRead(dep, FD, &buffer)) {
+      // Couldn't open the file now, so let's just assume the old path was
+      // canonical (enough).
+      realPath = dep;
+    } else {
+      realPath = buffer.str();
+      // Not much we can do about failing to close.
+      (void)close(FD);
+    }
+
+    // FIXME: do this decision more formally.
+    // Decide based on the extension of the raw dependency path, as the true
+    // file may have a different one.
+    auto ext = llvm::sys::path::extension(dep);
+    if (ext == SERIALIZED_MODULE_EXTENSION) {
+      swiftModules.push_back(realPath);
+    } else {
+      assert(ext == clang::api_notes::BINARY_APINOTES_EXTENSION);
```

Comment:
This is a bizarre thing to put in the trace file. apinotesc files are mostly generated on demand.

What about things in the list of dependencies that are neither modules nor apinotesc files?